src/watcher.rs

```diff
@@ -0,0 +1,316 @@
+use failure::{self, Fallible};
+use notify::{DebouncedEvent, RecursiveMode, Watcher};
+use std::collections::{HashMap, HashSet};
+use std::path::{Path, PathBuf};
+use std::sync::mpsc;
+use std::sync::{Arc, Mutex};
+use std::thread;
+use std::time::Duration;
+
+// FSWatch has an outgoing channel on which it pushes events for any of the files or directories it
+// is currently watching. Both directory and file watches are implemented using directory watches,
+// in order to be able to catch write-via-rename tricks like Vim does on save.
+//
+// Thus, FSWatch has a number of directory watches in place. These are all non-recursive watches;
+// recursive watches are connected directly to the FSWatch outgoing channel.
+// There is one thread, with entry point fswatch_service(), that collects all notify events and
+// filters out the ones that we're not interested in. Its incoming channel is attached to the
+// notify::RecommendedWatcher, and the outgoing channel is the outgoing channel of FSWatch.
+
+struct DirWatch {
+    /// whether to also watch the directory itself
+    full_directory: bool,
+    files: HashSet<String>,
+}
+
+fn interested(dirs: &Arc<Mutex<HashMap<PathBuf, DirWatch>>>, path: &Path) -> bool {
+    let dirs = dirs.lock().expect("FSWatch: mutex error");
+
+    if let Some(dw) = dirs.get(path) {
+        if dw.full_directory {
+            return true;
+        }
+    }
+
+    if let Some(parent) = path.parent() {
+        if let Some(dw) = dirs.get(parent) {
+            if let Some(name) = path.file_name().and_then(|n| n.to_str()) {
+                return dw.files.contains(name);
+            }
+        }
+    }
+
+    false
+}
+
+fn fswatch_service(
+    rx: mpsc::Receiver<DebouncedEvent>,
+    tx: mpsc::Sender<DebouncedEvent>,
+    dirs: Arc<Mutex<HashMap<PathBuf, DirWatch>>>,
+) -> Fallible<()> {
+    let mut notice_remove = HashSet::new();
+
+    // Policy:
+    // - If we're interested in a path, we handle and push through the event.
+    // - If we're not interested in a path:
+    //   - We _do_ remove the path from notice_remove if applicable; we may have been interested in
```

Comment:
Maybe I misunderstood this, but doesn't the remove happen in cases where we're interested as well? At least that's what I get from reading the match arms below.

Reply:
You are totally correct, we also remove the path from `notice_remove` in the "interested" case. I'll update the comment to better reflect reality.